File: rbac-manager-for-agents/internal/rbac/findings.go
// Copyright (C) ConfigHub, Inc.
// SPDX-License-Identifier: MIT

// RBAC hygiene analyzers, mirroring the published fleet-audit findings:
// wildcards, dangerous verbs, risky grants, cluster-admin bindings, orphaned
// bindings, and unbound ServiceAccounts. Pure functions over the snapshot;
// enforcement (gates) is server-side via Triggers — these are the analysis-only
// complement. Faithful Go port of the web app's findings.ts.

package rbac

import (
	"fmt"
	"slices"
	"sort"
	"strings"
)

// Severity ranks a finding's importance.
type Severity string

const (
	SeverityHigh   Severity = "high"
	SeverityMedium Severity = "medium"
	SeverityLow    Severity = "low"
)

// Finding is one hygiene issue discovered by an analyzer.
type Finding struct {
	ID           string         `json:"id"` // stable id for rendering and dedup
	Analyzer     string         `json:"analyzer"`
	Severity     Severity       `json:"severity"`
	Cluster      string         `json:"cluster"`
	Origin       ResourceOrigin `json:"origin"`
	ResourceKind string         `json:"resourceKind"`
	ResourceName string         `json:"resourceName"`
	Namespace    string         `json:"namespace,omitempty"`
	Message      string         `json:"message"`
}

var escalationVerbs = map[string]bool{"escalate": true, "bind": true, "impersonate": true}

func newFinding(analyzer string, severity Severity, origin ResourceOrigin, resourceKind, resourceName, message, namespace string) Finding {
	return Finding{
		ID:           fmt.Sprintf("%s:%s:%s:%s:%s", analyzer, origin.Cluster, resourceKind, namespace, resourceName),
		Analyzer:     analyzer,
		Severity:     severity,
		Cluster:      origin.Cluster,
		Origin:       origin,
		ResourceKind: resourceKind,
		ResourceName: resourceName,
		Namespace:    namespace,
		Message:      message,
	}
}

func wildcardFindings(cluster *ClusterRbac) []Finding {
	var out []Finding
	for _, role := range cluster.Roles {
		var parts []string
		for i, rule := range role.Rules {
			if slices.Contains(rule.Verbs, "*") {
				parts = append(parts, fmt.Sprintf("rule %d: wildcard verbs", i))
			}
			if slices.Contains(rule.Resources, "*") {
				parts = append(parts, fmt.Sprintf("rule %d: wildcard resources", i))
			}
			if slices.Contains(rule.APIGroups, "*") {
				parts = append(parts, fmt.Sprintf("rule %d: wildcard apiGroups", i))
			}
		}
		if len(parts) > 0 {
			out = append(out, newFinding("wildcard-rules", SeverityHigh, role.Origin, role.Kind, role.Name,
				fmt.Sprintf("Wildcard permissions (%s). Enumerate the specific verbs/resources needed.", strings.Join(parts, "; ")),
				role.Namespace))
		}
	}
	return out
}

func escalationFindings(cluster *ClusterRbac) []Finding {
	var out []Finding
	for _, role := range cluster.Roles {
		var verbs []string
		seen := map[string]bool{}
		for _, r := range role.Rules {
			for _, v := range r.Verbs {
				if escalationVerbs[v] && !seen[v] {
					seen[v] = true
					verbs = append(verbs, v)
				}
			}
		}
		if len(verbs) > 0 {
			out = append(out, newFinding("privilege-escalation-verbs", SeverityHigh, role.Origin, role.Kind, role.Name,
				fmt.Sprintf("Grants privilege-escalation verb(s): %s.", strings.Join(verbs, ", ")),
				role.Namespace))
		}
	}
	return out
}

var writeVerbs = map[string]bool{
	"create": true, "update": true, "patch": true, "delete": true, "deletecollection": true, "*": true,
}

// riskyGrantFindings flags grants that aren't wildcards but deserve eyes:
// secrets, exec, webhooks, CRDs.
func riskyGrantFindings(cluster *ClusterRbac) []Finding {
	var out []Finding
	for _, role := range cluster.Roles {
		var risks []string
		seen := map[string]bool{}
		add := func(r string) {
			if !seen[r] {
				seen[r] = true
				risks = append(risks, r)
			}
		}
		for _, rule := range role.Rules {
			writes := false
			for _, v := range rule.Verbs {
				if writeVerbs[v] {
					writes = true
					break
				}
			}
			if slices.Contains(rule.Resources, "secrets") {
				if writes {
					add("secrets write")
				} else {
					add("secrets read")
				}
			}
			if slices.Contains(rule.Resources, "pods/exec") || slices.Contains(rule.Resources, "pods/attach") {
				add("pod exec/attach")
			}
			if writes && (slices.Contains(rule.APIGroups, "admissionregistration.k8s.io") || slices.Contains(rule.APIGroups, "apiextensions.k8s.io")) {
				add("webhook/CRD write")
			}
		}
		if len(risks) > 0 {
			out = append(out, newFinding("risky-grants", SeverityMedium, role.Origin, role.Kind, role.Name,
				fmt.Sprintf("Sensitive access: %s.", strings.Join(risks, ", ")),
				role.Namespace))
		}
	}
	return out
}

func clusterAdminFindings(cluster *ClusterRbac) []Finding {
	var out []Finding
	for _, binding := range cluster.Bindings {
		isSuperuser := binding.RoleRef.Name == "cluster-admin"
		if !isSuperuser {
			if role := ResolveRoleRef(binding, cluster); role != nil {
				for _, r := range EffectiveRules(role, cluster) {
					if slices.Contains(r.Verbs, "*") && slices.Contains(r.Resources, "*") && slices.Contains(r.APIGroups, "*") {
						isSuperuser = true
						break
					}
				}
			}
		}
		if !isSuperuser {
			continue
		}
		subjects := "(no subjects)"
		if len(binding.Subjects) > 0 {
			keys := make([]string, len(binding.Subjects))
			for i, s := range binding.Subjects {
				keys[i] = SubjectKey(s)
			}
			subjects = strings.Join(keys, ", ")
		}
		severity := SeverityMedium
		if binding.Kind == "ClusterRoleBinding" {
			severity = SeverityHigh
		}
		out = append(out, newFinding("cluster-admin-bindings", severity, binding.Origin, binding.Kind, binding.Name,
			fmt.Sprintf("Grants superuser (%s) to: %s.", binding.RoleRef.Name, subjects),
			binding.Namespace))
	}
	return out
}

func orphanedBindingFindings(cluster *ClusterRbac) []Finding {
	var out []Finding
	for _, binding := range cluster.Bindings {
		if ResolveRoleRef(binding, cluster) != nil {
			continue
		}
		if IsBuiltinRoleName(binding.RoleRef.Name) {
			continue
		}
		out = append(out, newFinding("orphaned-bindings", SeverityMedium, binding.Origin, binding.Kind, binding.Name,
			fmt.Sprintf("References %s %q, which does not exist on this cluster. Remove the binding or restore the role.", binding.RoleRef.Kind, binding.RoleRef.Name),
			binding.Namespace))
	}
	return out
}

func unboundServiceAccountFindings(cluster *ClusterRbac) []Finding {
	bound := map[string]bool{}
	for _, binding := range cluster.Bindings {
		for _, s := range binding.Subjects {
			if s.Kind == "ServiceAccount" {
				bound[s.Namespace+"/"+s.Name] = true
			}
		}
	}
	var out []Finding
	for _, sa := range cluster.ServiceAccounts {
		if bound[sa.Namespace+"/"+sa.Name] {
			continue
		}
		out = append(out, newFinding("unbound-service-accounts", SeverityLow, sa.Origin, "ServiceAccount", sa.Name,
			"ServiceAccount has no role bindings in this snapshot — possibly unused.",
			sa.Namespace))
	}
	return out
}

var analyzers = []func(*ClusterRbac) []Finding{
	wildcardFindings,
	escalationFindings,
	riskyGrantFindings,
	clusterAdminFindings,
	orphanedBindingFindings,
	unboundServiceAccountFindings,
}

var severityOrder = map[Severity]int{SeverityHigh: 0, SeverityMedium: 1, SeverityLow: 2}

// AnalyzeFleet runs every analyzer over every cluster, sorted by severity, then
// cluster, then id.
func AnalyzeFleet(clusters map[string]*ClusterRbac) []Finding {
	var out []Finding
	for _, cluster := range clusters {
		for _, analyzer := range analyzers {
			out = append(out, analyzer(cluster)...)
		}
	}
	sort.SliceStable(out, func(i, j int) bool {
		if severityOrder[out[i].Severity] != severityOrder[out[j].Severity] {
			return severityOrder[out[i].Severity] < severityOrder[out[j].Severity]
		}
		if out[i].Cluster != out[j].Cluster {
			return out[i].Cluster < out[j].Cluster
		}
		return out[i].ID < out[j].ID
	})
	return out
}

File: fleet-ql/src/rbac/findings.ts
// RBAC hygiene analyzers, mirroring the published fleet-audit findings:
// wildcards, dangerous verbs, risky grants, cluster-admin bindings, orphaned
// bindings, and unbound ServiceAccounts. Pure functions over the snapshot;
// enforcement (gates) is server-side via Triggers — these are the
// analysis-only complement.

import { ClusterRbac, ResourceOrigin, subjectKey } from './model';
import { effectiveRules, isBuiltinRoleName, resolveRoleRef } from './semantics';

export type Severity = 'high' | 'medium' | 'low';

export interface Finding {
  /** Stable id for list rendering and dedup. */
  id: string;
  analyzer: string;
  severity: Severity;
  cluster: string;
  origin: ResourceOrigin;
  resourceKind: string;
  resourceName: string;
  namespace?: string;
  message: string;
}

const ESCALATION_VERBS = new Set(['escalate', 'bind', 'impersonate']);

function finding(
  analyzer: string,
  severity: Severity,
  origin: ResourceOrigin,
  resourceKind: string,
  resourceName: string,
  message: string,
  namespace?: string,
): Finding {
  return {
    id: `${analyzer}:${origin.cluster}:${resourceKind}:${namespace ?? ''}:${resourceName}`,
    analyzer,
    severity,
    cluster: origin.cluster,
    origin,
    resourceKind,
    resourceName,
    namespace,
    message,
  };
}

function wildcardFindings(cluster: ClusterRbac): Finding[] {
  const out: Finding[] = [];
  for (const role of cluster.roles) {
    const parts: string[] = [];
    role.rules.forEach((rule, i) => {
      if (rule.verbs.includes('*')) parts.push(`rule ${i}: wildcard verbs`);
      if (rule.resources.includes('*')) parts.push(`rule ${i}: wildcard resources`);
      if (rule.apiGroups.includes('*')) parts.push(`rule ${i}: wildcard apiGroups`);
    });
    if (parts.length > 0) {
      out.push(
        finding(
          'wildcard-rules',
          'high',
          role.origin,
          role.kind,
          role.name,
          `Wildcard permissions (${parts.join('; ')}). Enumerate the specific verbs/resources needed.`,
          role.namespace,
        ),
      );
    }
  }
  return out;
}

function escalationFindings(cluster: ClusterRbac): Finding[] {
  const out: Finding[] = [];
  for (const role of cluster.roles) {
    const verbs = new Set(
      role.rules.flatMap((r) => r.verbs.filter((v) => ESCALATION_VERBS.has(v))),
    );
    if (verbs.size > 0) {
      out.push(
        finding(
          'privilege-escalation-verbs',
          'high',
          role.origin,
          role.kind,
          role.name,
          `Grants privilege-escalation verb(s): ${[...verbs].join(', ')}.`,
          role.namespace,
        ),
      );
    }
  }
  return out;
}

/** Grants that aren't wildcards but deserve eyes: secrets, exec, webhooks, CRDs. */
function riskyGrantFindings(cluster: ClusterRbac): Finding[] {
  const out: Finding[] = [];
  const WRITE_VERBS = ['create', 'update', 'patch', 'delete', 'deletecollection', '*'];
  for (const role of cluster.roles) {
    const risks: string[] = [];
    for (const rule of role.rules) {
      const writes = rule.verbs.some((v) => WRITE_VERBS.includes(v));
      if (rule.resources.some((r) => r === 'secrets')) {
        risks.push(writes ? 'secrets write' : 'secrets read');
      }
      if (rule.resources.some((r) => r === 'pods/exec' || r === 'pods/attach')) {
        risks.push('pod exec/attach');
      }
      if (
        writes &&
        rule.apiGroups.some((g) => g === 'admissionregistration.k8s.io' || g === 'apiextensions.k8s.io')
      ) {
        risks.push('webhook/CRD write');
      }
    }
    if (risks.length > 0) {
      out.push(
        finding(
          'risky-grants',
          'medium',
          role.origin,
          role.kind,
          role.name,
          `Sensitive access: ${[...new Set(risks)].join(', ')}.`,
          role.namespace,
        ),
      );
    }
  }
  return out;
}

function clusterAdminFindings(cluster: ClusterRbac): Finding[] {
  const out: Finding[] = [];
  for (const binding of cluster.bindings) {
    let isSuperuser = binding.roleRef.name === 'cluster-admin';
    if (!isSuperuser) {
      const role = resolveRoleRef(binding, cluster);
      if (role) {
        isSuperuser = effectiveRules(role, cluster).some(
          (r) => r.verbs.includes('*') && r.resources.includes('*') && r.apiGroups.includes('*'),
        );
      }
    }
    if (isSuperuser) {
      const subjects = binding.subjects.map(subjectKey).join(', ') || '(no subjects)';
      out.push(
        finding(
          'cluster-admin-bindings',
          binding.kind === 'ClusterRoleBinding' ? 'high' : 'medium',
          binding.origin,
          binding.kind,
          binding.name,
          `Grants superuser (${binding.roleRef.name}) to: ${subjects}.`,
          binding.namespace,
        ),
      );
    }
  }
  return out;
}

function orphanedBindingFindings(cluster: ClusterRbac): Finding[] {
  const out: Finding[] = [];
  for (const binding of cluster.bindings) {
    if (resolveRoleRef(binding, cluster)) continue;
    if (isBuiltinRoleName(binding.roleRef.name)) continue;
    out.push(
      finding(
        'orphaned-bindings',
        'medium',
        binding.origin,
        binding.kind,
        binding.name,
        `References ${binding.roleRef.kind} "${binding.roleRef.name}", which does not exist on this cluster. Remove the binding or restore the role.`,
        binding.namespace,
      ),
    );
  }
  return out;
}

function unboundServiceAccountFindings(cluster: ClusterRbac): Finding[] {
  const bound = new Set<string>();
  for (const binding of cluster.bindings) {
    for (const s of binding.subjects) {
      if (s.kind === 'ServiceAccount') bound.add(`${s.namespace ?? ''}/${s.name}`);
    }
  }
  const out: Finding[] = [];
  for (const sa of cluster.serviceAccounts) {
    if (bound.has(`${sa.namespace}/${sa.name}`)) continue;
    out.push(
      finding(
        'unbound-service-accounts',
        'low',
        sa.origin,
        'ServiceAccount',
        sa.name,
        `ServiceAccount has no role bindings in this snapshot — possibly unused.`,
        sa.namespace,
      ),
    );
  }
  return out;
}

const ANALYZERS: ((cluster: ClusterRbac) => Finding[])[] = [
  wildcardFindings,
  escalationFindings,
  riskyGrantFindings,
  clusterAdminFindings,
  orphanedBindingFindings,
  unboundServiceAccountFindings,
];

/** Run every analyzer over every cluster. */
export function analyzeFleet(clusters: Map<string, ClusterRbac>): Finding[] {
  const out: Finding[] = [];
  for (const cluster of clusters.values()) {
    for (const analyzer of ANALYZERS) {
      out.push(...analyzer(cluster));
    }
  }
  const severityOrder: Record<Severity, number> = { high: 0, medium: 1, low: 2 };
  return out.sort(
    (a, b) =>
      severityOrder[a.severity] - severityOrder[b.severity] ||
      a.cluster.localeCompare(b.cluster) ||
      a.id.localeCompare(b.id),
  );
}
